File: OIL/CPP/src/interpreter/Interpreter.cpp
/*
** ZIPLEIX PROJECT, 2024
** OIL [WSL : Ubuntu]
** File description:
** Interpreter
*/

#include "Interpreter.hpp"

Interpreter::Interpreter(std::vector<std::any> tokens, std::map<std::string, unsigned int> etiquetteTracker, size_t stackSize): _tokens(tokens), _etiquetteTracker(etiquetteTracker), _pc(0), _stack(stackSize)
{
    _opcodes["HALT"] = std::bind(&Interpreter::halt, this);
    _opcodes["PUSH"] = std::bind(&Interpreter::push, this);
    _opcodes["POP"] = std::bind(&Interpreter::pop, this);
    _opcodes["ADD"] = std::bind(&Interpreter::add, this);
    _opcodes["SUB"] = std::bind(&Interpreter::sub, this);
    _opcodes["PRINT"] = std::bind(&Interpreter::print, this);
    _opcodes["READ"] = std::bind(&Interpreter::read, this);
    _opcodes["JUMP.EQ.0"] = std::bind(&Interpreter::jumpEq0, this);
    _opcodes["JUMP.GT.0"] = std::bind(&Interpreter::jumpGt0, this);
}

void Interpreter::interprete() {
    if (_tokens.empty()) {
        std::cerr << "Error: No tokens to interprete" << std::endl;
        return;
    }

    while (1) {
        if (_pc >= _tokens.size()) {
            break;
        }
        std::string opcode = std::any_cast<std::string>(_tokens[_pc]);
        _pc++;

        if (_opcodes.find(opcode) == _opcodes.end()) {
            std::cerr << "Error: Unknown opcode " << opcode << std::endl;
            return;
        } else {
            if (_opcodes[opcode]()) {
                break;
            }
        }
    }
}

char Interpreter::halt() {
    return 1;
}

char Interpreter::push() {
    int value = std::any_cast<int>(_tokens[_pc]);
    _pc++;
    _stack.push(value);
    return 0;
}

char Interpreter::pop() {
    _stack.pop();
    return 0;
}

char Interpreter::add() {
    int a = _stack.top();
    _stack.pop();
    int b = _stack.top();
    _stack.pop();
    _stack.push(a + b);
    return 0;
}

char Interpreter::sub() {
    int a = _stack.top();
    _stack.pop();
    int b = _stack.top();
    _stack.pop();
    _stack.push(b - a);
    return 0;
}

char Interpreter::print() {
    std::string string_literal = std::any_cast<std::string>(_tokens[_pc]);
    _pc++;
    std::cout << string_literal << std::endl;
    return 0;
}

char Interpreter::read() {
    int value;
    std::cin >> value;
    _stack.push(value);
    return 0;
}

char Interpreter::jumpEq0() {
    std::string etiquette = std::any_cast<std::string>(_tokens[_pc]);
    if (_stack.top() == 0) {
        _pc = _etiquetteTracker[etiquette];
    } else {
        _pc++;
    }
    return 0;
}

char Interpreter::jumpGt0() {
    std::string etiquette = std::any_cast<std::string>(_tokens[_pc]);
    if (_stack.top() > 0) {
        _pc = _etiquetteTracker[etiquette];
    } else {
        _pc++;
    }
    return 0;
}

File: OIL/V1/python/interpreter.py
import sys

# read the name of the script file
script_path = sys.argv[1]

#############
# TOKENISER #
#############

# read line one by one
script_lines = []
with open(script_path, 'r') as script_file:
    script_lines = [line.strip() for line in script_file.readlines()]

tokens = []
token_counter = 0
etiquette_tracker = {}
for line in script_lines:
    parts = line.split(' ')
    opcode = parts[0]

    # check empty line
    if len(opcode) == 0:
        continue

    # check if it is an etiquette
    if opcode[-1] == ':':
        etiquette_tracker[opcode[:-1]] = token_counter
        continue

    # store opcode token
    tokens.append(opcode)
    token_counter += 1

    # handle each opcode
    if opcode == 'PUSH':
        # need a number after PUSH
        number = int(parts[1])
        tokens.append(number)
        token_counter += 1
    elif opcode == 'PRINT':
        # parse string literal
        string_literal = ' '.join(parts[1:])[1:-1]
        tokens.append(string_literal)
        token_counter += 1
    elif opcode == 'JUMP.EQ.0':
        etiquette = parts[1]
        tokens.append(etiquette)
        token_counter += 1
    elif opcode == 'JUMP.GT.0':
        etiquette = parts[1]
        tokens.append(etiquette)
        token_counter += 1

###############
# INTERPRETER #
###############

class Stack:
    def __init__(self, size: int):
        self.buf = [0 for _ in range(size)]
        self.sp = -1

    def push(self, value: int):
        self.sp += 1
        self.buf[self.sp] = value

    def pop(self) -> int:
        value = self.buf[self.sp]
        self.sp -= 1
        return value

    def top(self) -> int:
        return self.buf[self.sp]

    def print(self):
        print(self.buf[:self.sp + 1])

pc = 0
stack = Stack(256)

while tokens[pc] != 'HALT':
    opcode = tokens[pc]
    pc += 1

    if opcode == 'PUSH':
        value = tokens[pc]
        pc += 1
        stack.push(value)
    elif opcode == 'POP':
        stack.pop()
    elif opcode == 'ADD':
        value1 = stack.pop()
        value2 = stack.pop()
        stack.push(value1 + value2)
    elif opcode == 'SUB':
        value1 = stack.pop()
        value2 = stack.pop()
        stack.push(value2 - value1)
    elif opcode == 'PRINT':
        value = tokens[pc]
        pc += 1
        print(value)
    elif opcode == 'READ':
        value = int(input())
        stack.push(value)
    elif opcode == 'JUMP.EQ.0':
        number = stack.top()
        if number == 0:
            pc = etiquette_tracker[tokens[pc]]
        else:
            pc += 1
    elif opcode == 'JUMP.GT.0':
        number = stack.top()
        if number > 0:
            pc = etiquette_tracker[tokens[pc]]
        else:
            pc += 1
    else:
        raise Exception(f'Unknown opcode: {opcode}')

